### PR TITLE
[release_tool] QA-178: Implement --hosted-release flow

### PR DESCRIPTION
**Pull request by lluiscampos**
This workflow will allow the user to generate the hosted Mender
production tags based on current versions in staging.

See README for more details.

### CODE REVIEW

**Comment by lluiscampos** [comment]
Stopped the integration pipeline.

File: extra/release_tool.py [hunk 1]
```diff
@@ -2720,6 +2727,130 @@ def do_map_name(args):
...
+        print(
+            "do_hosted_release can only be called from staging branch; current branch is %s"
+            % ref
+        )
+        sys.exit(2)
+
+    # Recreate state dict for the function helpers
+    state = {}
+    state["integration"] = {}
+    state["integration"]["following"] = "staging"
```
[the remark this answers is not in the document]
Comment (lluiscampos) [reply extra/release_tool.py:2784]: Yep. As you have seen, I decided not to use the file at all. This is something I would like to discuss, so thanks for bringing it up.

This means that if you abort you need to start again (even typing again which dir has all repos) but also ensures that you can not tag something that is not latest in `staging` branches.

My reasoning is that this release flow has no knowledge of "builds" (or tests, for all that matters), so the only allowed versions to be released must always be latest `staging` (i.e. whatever is right now in HM Staging and we have tested there). Then, to be strict with that principle, I didn't want to have a state saved on disc that you can resume later. Does it make sense?

Do you @tranchitella have an opinion on this?

[the remark this answers is not in the document]
Comment (kacf) [reply extra/release_tool.py:2784]: > This means that if you abort you need to start again (even typing again which dir has all repos) but also ensures that you can not tag something that is not latest in `staging` branches.

But this information is not recorded in the state file. The state file only stores which branches you want to follow, not what is in them. On the contrary, the actual branch content is figured out on each run, using the `tag_avail` mechanism. Even the build tags are figured out this way, on the fly.

So the main reason for not storing anything is just that the information is trivial, IOW, there won't ever be anything other than `staging` in there. For this simple usage, without builds, this approach probably makes sense though. This release mode is not menu/loop based like the on-prem releases, so every invocation is a one-shot.

[the remark this answers is not in the document]
Comment (lluiscampos) [reply extra/release_tool.py:2784]: > But this information is not recorded in the state file. (...)

Correct! I got it wrong. Thanks.

> So the main reason for not storing anything is just that the information is trivial, IOW, there won't ever be anything other than `staging` in there. For this simple usage, without builds, this approach probably makes sense though. This release mode is not menu/loop based like the on-prem releases, so every invocation is a one-shot.

Yeah, the only thing I can see worth saving is the location of the repos. But then we need to maintain two different files (regular release, hosted mender releases) or ask the user if we wants to continue the previous one (I think this won't make sense).

@kacf To conclude the topic: do you then support the idea of not having state file for this type of release?

**Pull request merged**